text/0000-code-literals.md

```diff
@@ -0,0 +1,417 @@
+    ```rust
+    let _ = ```
+        some code
+        ```;
+    ```
+
+  - **3+N double-quotes**
+    ```rust
+    let _ = """
+        some code
+        """;
+    ```
+
+  - **3+N single quotes**
+    ```rust
+    let _ = '''
+        some code
+        ''';
+    ```
+
+  - **Word prefix + N+1 hashes**
+    ```rust
+    let _ = code#"
+        some code
+        "#;
+    ```
+
+  - **Single character prefix + N+1 hashes**
+    ```rust
+    let _ = m#"
+        some code
+        "#;
+    ```
+    (note: `c` is already reserved for C strings)
+
+### Indentation rules
+
+  - :heavy_check_mark: **Relative to closing quote + retain final newline**
+
+    Benefits:
+      - Allows every possible indentation to be represented.
+      - Simple rule.
+      - The value of the string is obvious and intuitive.
+
+    Drawbacks:
+      - It is not possible to represent strings without a trailing newline.
+
+  - **Relative to closing quote + remove final newline**
+
+    Benefits:
+      - Allows every possible indentation to be represented.
+      - Simple rule.
+      - Strings without a final newline can be represented.
+
+    Drawbacks:
+      - There are two ways to represent the empty string.
+      - It is unintuitive that two empty lines between quotes results in
+        a single newline.
+
+  - **Relative to first non-empty line**
+
+    Benefits:
+      - Simple rule.
+      - The value of the string is obvious and intuitive.
+      - Strings without a final newline can be represented.
+
+    Drawbacks:
+      - Some indentations cannot be represented.
+
+  - **Relative to least indented line**
+
+    Benefits:
+      - Simple rule.
+      - The value of the string is obvious and intuitive.
+      - Strings without a final newline can be represented.
+
+    Drawbacks:
+      - Some indentations cannot be represented.
+
+### Modifications
+
+  - :heavy_check_mark: **Language hint directly following opening quote**
+
+    This is intended to allow extra information (eg. language) to be
+    conveyed by the programmer to macros and/or their IDE. For example:
+    ```rust
+    let _ = ```sql
+        SELECT * FROM table;
+        ```;
+    ```
+    Here, an intelligent IDE could apply syntax highlighting to the nested
+    code block, knowing that the code is SQL.
+
+  - :heavy_check_mark: **Annotation on closing quote to remove trailing
+    newline**
+
+    For indentation rules where the final quote must appear on
+    its own line and there is no way to represent a string without
+    a trailing newline, a modification character could be used.
+
+    For example:
+    ```rust
+    let _ = ```
+        no trailing newline
+        !```;
+    ```
+
+    Or (the less serious suggestion of)...
+    ```rust
+    let _ = ```
+        no trailing newline
+        🚫```;
+    ```
+
+    This could be used with any quote style and is unambiguous because
+    nothing can otherwise appear on the same line prior to the closing
+    quote.
+
+# Prior art
```

Comment:
Python's [`cleandoc`](https://docs.python.org/3/library/inspect.html#inspect.cleandoc) is often used for this purpose, and could set a good example of "expected behavior"